ouroboros-consensus/docs/HardFork.md

```diff
@@ -0,0 +1,211 @@
+
+   Notably, endorsements for proposals that are not yet stably-confirmed (or do
+   not even exist) are not invalid but rather silently ignored. In other words,
+   no validation applies to the `headerProtocolVersion` field.
+
+5. Once the number of endorsers satisfies a threshold (same as for voting), the
+   confirmed proposal becomes a _candidate_ proposal.
+
+6. _At the beginning of an epoch_, the candidate proposal with the greatest
+   protocol version number among those candidates whose threshold-satisfying
+   endorsement is stable (ie the block is `>=2k` slots old) is _adopted_: the
+   new protocol parameter values have now been changed.
+
+   If there was no stably-candidated proposal, then nothing happens. Everything
+   is retained; in particular, a candidate proposal whose threshold-satisfying
+   endorsement was not yet stable will be adopted at the subsequent epoch unless
+   it is surpassed in the meantime.
+
+   When a candidate is adopted, all record of other proposals/votes/endorsements
+   -- regardless of their state -- is discarded. The explanation for this is
+   that such proposals would now be interpreted as an update to the newly
+   adopted parameter values, whereas they were validated as an update to the
+   previously adopted parameter values.
+
+In summary, the following diagram tracks the progress of a proposal that's
+eventually adopted. For other proposals, the path short circuits to a
+"rejected/discarded" status at some point.
+
+```
+active proposal
+    --> (sufficient votes)
+confirmed proposal
+    --> (2k slots later)
+stably-confirmed proposal
+    --> (sufficient endorsements)
+candidate proposal
+   --> (2k slots later)
+stably-candidated proposal    (Frisby: stably-nominated?)
+   --> (epoch transition)
+adopted proposal
+```
+
+### Initiating the hard fork
+
+Proposals to initiate the hard fork can be submitted and voted on before all
+core nodes are ready. After all, once a proposal is "stably-confirmed", it will
+effectively remain so indefinitely until nodes endorse it (or it gets superseded
+by another proposal). This means that nodes can vote to initiate the hard fork,
+_then_ wait for everybody to update their software, and once updated, the
+proposal is endorsed and eventually the hard fork is initiated.
+
+Endorsement is somewhat implicit. The node operator does not submit an explicit
+"endorsement transaction", but instead restarts the node (probably after a
+software update that makes the node ready to support the hard fork) with a new
+protocol version (as part of a config file or command line parameter), which
+then gets included in the blocks that the node produces (this value is part of
+the static `ByronConfig`: `byronProtocolVersion`).
+
+(Note that a node restart is necessary for _any_ change to a protocol parameter,
+even though most parameters do not require any change to the software at all.)
+
+### Software version
```

Comment:
Maybe rename this to `Header software version` lest people will think we ignore the software version in the spec, whereas we only ignore the software version in the block header.